notebooks/examples/python/BritishNationalGrid.py

```diff
@@ -0,0 +1,370 @@
+# MAGIC Since we are operating in EPSG:27700 and using BNG as our indexing system, we will use the location data provided via Northings and Eastings coordinates.
+
+# COMMAND ----------
+
+uprns_table = (
+  uprns_table
+    .withColumn("uprn_point", mos.st_point(col("X_COORDINATE"), col("Y_COORDINATE")))
+    # we are using WKT here for simpler displaying, use WKB for faster query run time
+    .withColumn("uprn_point", mos.st_aswkt("uprn_point")) 
+    .where(mos.st_hasvalidcoordinates("uprn_point", lit('EPSG:27700'), lit('reprojected_bounds')))
+    .where(mos.st_isvalid(col("uprn_point")))
+    .drop("LATITUDE", "LONGITUDE")
+)
+display(uprns_table)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Next step is optional. Howerver, since we are constructing POINT geometries and ensuring they are valid it is prudent to write out the validated dataset. <br>
+# MAGIC That way we are making sure validation is performed only once ate ingestion and not each time spark runs the queries (due to spark lazy evaluation). 
+
+# COMMAND ----------
+
+uprns_table.write.format("delta").mode("overwrite").saveAsTable("uprns_bng_table")
+
+# COMMAND ----------
+
+uprns_table = spark.read.table("uprns_bng_table")
+
+# COMMAND ----------
+
+uprns_table.count()
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Spatial Joins
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC We can use Mosaic to perform spatial joins both with and without Mosaic indexing strategies. </br>
+# MAGIC Indexing is very important when handling very different geometries both in size and in shape (ie. number of vertices). </br>
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Getting the optimal resolution
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC We can use Mosaic functionality to identify how to best index our data based on the data inside the specific dataframe. </br>
+# MAGIC Selecting an apropriate indexing resolution can have a considerable impact on the performance. </br>
+
+# COMMAND ----------
+
+from mosaic import MosaicFrame
+
+postcodes_mosaic_frame = MosaicFrame(postcodes, "geometry")
+optimal_resolution = postcodes_mosaic_frame.get_optimal_resolution(sample_fraction=0.75)
+optimal_resolution_str = postcodes_mosaic_frame.get_optimal_resolution_str(sample_fraction=0.75)
+
+print(f"""
+  Optimal resolution code is :{optimal_resolution}.
+  Optimal resolution name is :{optimal_resolution_str}.
+""")
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Not every resolution will yield performance improvements. </br>
+# MAGIC By a rule of thumb it is always better to under-index than over-index - if not sure select a lower resolution. </br>
+# MAGIC Higher resolutions are needed when we have very imballanced geometries with respect to their size or with respect to the number of vertices. </br>
+# MAGIC In such case indexing with more indices will considerably increase the parallel nature of the operations. </br>
```

Comment:
indexing -> tessellating
more indices -> higher resolution
increase the parallel nature of the operations -> increase the parallel nature of the downstream operations

Reply:
reworded